app/domain/authentication/authn_oidc/authenticator.rb

```diff
@@ -0,0 +1,84 @@
+module Authentication
+  module AuthnOidc
+    class AuthenticationError < RuntimeError; end
+    class NotFoundError < RuntimeError; end
+    class OIDCConfigurationError < RuntimeError; end
+    class OIDCAuthenticationError < RuntimeError; end
+
+    class Authenticator
+
+      def initialize(env:)
+        @env = env
+      end
+
+      def valid?(input)
+        @authenticator_name = input.authenticator_name
+        @service_id = input.service_id
+        @conjur_account = input.account
+        @request_body = input.request.body.read
+
+        verify_service_enabled
+
+        oidc_authn_service = AuthenticationService.new(service.identifier, conjur_account)
+        id_token, user_info = oidc_authn_service.get_user_details(request_body)
+
+        # TODO: validate id_token - if not raise error
+
+        validate_user_info(user_info, id_token.sub)
+
+        username = user_info.preferred_username
+        input.instance_variable_set(:@username, username)
+
+        true
+      end
+
+      private
+
+      def authenticator_name
```

Comment:
Is there a reason you need these four getters? I would suggest removing them and just referring to the instance variable directly w/ `@`.

Reply:
done